### DISCUSSION TITLE
Interface for adding custom costs in the trajectory optimization/ik/motion planning

### DISCUSSION THREAD

**Discussion by waldezjr** [NVlabs/curobo#80] [Software Q&A] [answered]
Within the library, several classes config the current cost functions, and as I understand, there is an easy way of changing most parameters through the yml files in /content folder such as gradient_trajopt.yml files.

However, the examples don't seem to be clear on how is it possible to implement custom Cost functions, with custom parameters, etc.

1) Is it the case that one would need to implement a class deriving from the CostBase class?
1.2) If we take this route, is there any other interface/implementation that needs to be done elsewhere?

2) More specifically, I would like to try to enforce trajectories with similar (not necessarily constrained) end-effector orientations throughout the entire planned trajectory. 
Has anyone tried something similar? Or would you have any suggestions for modifying the current code for this matter?

**Comment by balakumar-s** [comment, marked as answer]
There are two ways to add your own cost function:
1. Add them to the [ArmReacher](https://github.com/NVlabs/curobo/blob/256227b0b6ce2f06892aef24bf7d933bb017806a/src/curobo/rollout/arm_reacher.py#L225) class. This will allow you to use the existing IK, TrajOpt, and MotionGen wrappers implemented in https://github.com/NVlabs/curobo/tree/main/src/curobo/wrap/reacher.
2. Create a new rollout class that inherits ArmReacher or ArmBase and implement your costs in a cost function. This would also require you to create your own wrappers for IK/TrajOpt/MotionGen if you plan on using them.

In both these cases, you have to use CostBase as the base class. 

If you use option (1), then you can modify this function: https://github.com/NVlabs/curobo/blob/256227b0b6ce2f06892aef24bf7d933bb017806a/src/curobo/rollout/arm_reacher.py#L89 to also load your config parameters for the custom cost function. 

For your second question on enforcing orientation along the trajectory:
We have this implemented. Change this scalar to a value less than 1: https://github.com/NVlabs/curobo/blob/256227b0b6ce2f06892aef24bf7d933bb017806a/src/curobo/content/configs/task/gradient_trajopt.yml#L38 and then make the last three scalars 0.0 in this line : https://github.com/NVlabs/curobo/blob/256227b0b6ce2f06892aef24bf7d933bb017806a/src/curobo/content/configs/task/gradient_trajopt.yml#L34
You also have to do it in this file: https://github.com/NVlabs/curobo/blob/main/src/curobo/content/configs/task/finetune_trajopt.yml

The `run_weight` scalar scales the pose cost at every time step and the `run_vec_weight` allows you to scale the individual axes.

**Comment by balakumar-s** [comment]
We have added constrained planning which enables locking orientations. See here for examples: https://curobo.org/source/advanced_examples/3_constrained_planning.html
